### PR TITLE
Contract types as mapping keys.

### LINKED ISSUE #8128: Support contracts and interfaces for mapping keys.
## Abstract
Contracts and interfaces should be allowed for mapping keys.
```
interface IApple {}
contract Banana {}
interface ICherry {}

contract Apple {
    mapping(IApple => bool) public apples;
    mapping(Banana => bool) public bananas;
    mapping(address => bool) public cherries;

    constructor(ICherry cherry) public {
        cherries[cherry] = true;
    }
}
```
> ParserError: Expected elementary type name for mapping key type
> `mapping(IApple => bool) public apples;`

> TypeError: Type contract ICherry is not implicitly convertible to expected type address.
> `cherries[cherry] = true;`

## Motivation
When creating a mapping of contract to thing I have to use the elementary types as the mapping key which means `address` in this case.  Unfortunately, since contracts and interfaces do not implicitly cast to `address`, this means every time I use the interface I must cast to address first.  This leads to a lot of unnecessary code, and can sometimes leads to bug as explicit casts are error prone.

## Specification
If a contract or interface is used as a `mapping` key, it should be functionally equivalent as an `address` being used as the mapping key with the exception of type checking when the mapping is indexed.

## Backwards Compatibility
None.

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #8128 
Am I missing something? This is basically it, isn't it? I'm not sure why this was assigned 5 points...

### CODE REVIEW

File: docs/types/mapping-types.rst [hunk 1]
```diff
@@ -7,9 +7,8 @@ Mapping Types
 Mapping types use the syntax ``mapping(_KeyType => _ValueType)`` and variables
 of mapping type are declared using the syntax ``mapping(_KeyType => _ValueType) _VariableName``.
 The ``_KeyType`` can be any
-built-in value type plus ``bytes`` and ``string``. User-defined
-or complex types such as contract types, enums, mappings, structs or array types
-apart from ``bytes`` and ``string`` are not allowed.
+built-in value type, ``bytes``, ``string`` or any contract type. Other user-defined
```
Comment (chriseth) [inline docs/types/mapping-types.rst]: We actually wanted to include enums.

Comment (ekpyron) [reply docs/types/mapping-types.rst]: Ok, I can do that...

**Comment by chriseth** [comment]
Please also implement getter functions.

**Comment by ekpyron** [comment]
I also missed AST importing.

**Comment by ekpyron** [comment]
> Please also implement getter functions.

They are already generated, but I'll add explicit tests...

File: test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        table[k] = v;
+    }
+    function get(E k) public returns (uint8) {
+        return this.table(k);
+    }
+}
+// ----
+// table(uint8): 0 -> 0
+// table(uint8): 0x01 -> 0
+// table(uint8): 0xa7 -> 0
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol:16]: Do we want these to revert with the old encoder as well?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `8fc8dbe675f1d4ed410e94be0a4c44d6c07b7481`:
```
Coding style error:
 libsolidity/codegen/ir/IRGenerator.cpp:190: } while((mappingType = dynamic_cast<MappingType const*>(mappingType->valueType())));

```
Please check that your changes are working as intended.

**Comment by ekpyron** [comment]
I also added a last commit with some Yul IR code generation for mapping getters, but I'm not sure about it and we might want to postpone that part and remove it from this PR...

File: docs/types/mapping-types.rst (hunk 1 above)
Comment (chriseth) [reply docs/types/mapping-types.rst]: Please change the documentation accordingly.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -2876,6 +2876,25 @@ void TypeChecker::endVisit(Literal const& _literal)
 	_literal.annotation().isPure = true;
 }
 
+bool TypeChecker::visit(Mapping const& _mapping)
+{
+	if (auto const* keyType = dynamic_cast<UserDefinedTypeName const*>(&_mapping.keyType()))
+	{
+		if (
+			!dynamic_cast<ContractDefinition const*>(keyType->annotation().referencedDeclaration) &&
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Maybe better to use `keyType->annotation().type->category()`?

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 4]
```diff
@@ -159,21 +159,70 @@ string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
 	solAssert(!_varDecl.isConstant(), "");
 	solAssert(_varDecl.isStateVariable(), "");
 
-	solUnimplementedAssert(type->isValueType(), "");
+	if (auto const* mappingType = dynamic_cast<MappingType const*>(type))
+		return m_context.functionCollector()->createFunction(functionName, [&]() {
+			pair<u256, unsigned> slot_offset = m_context.storageLocationOfVariable(_varDecl);
+			solAssert(slot_offset.second == 0, "");
+			FunctionType funType(_varDecl);
+			solAssert(funType.returnParameterTypes().size() == 1, "");
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp]:
```suggestion
			solUnimplementedAssert(funType.returnParameterTypes().size() == 1, "");
```
?

**Comment by ekpyron** [comment]
Now the fact that the enum getters seem to throw on CI, but not for me locally is a bit creepy...

File: test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol:16]: I think the old encoder already enforces enums in abi decoding, doesn't it?

**Comment by ekpyron** [comment]
> Now the fact that the enum getters seem to throw on CI, but not for me locally is a bit creepy...

Not so creepy after all - it's just the ABIEncoderV2 runs, that makes sense.

File: test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol (hunk 2 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol:16]: Nope, it doesn't...
https://github.com/ethereum/solidity/blob/7a194ffdabb50dff3b29a2973255ec05b2b3960e/test/libsolidity/ABIDecoderTests.cpp#L90

Comment (ekpyron) [reply test/libsolidity/semanticTests/types/mapping_enum_key_getter.sol:16]: Not sure what to do about that... we can't really change the decoder in general without it being strictly speaking breaking... and making an exception in the decoder *only* for these mapping getters would be rather hacky as well...

File: docs/types/mapping-types.rst (hunk 1 above)
Comment (ekpyron) [reply docs/types/mapping-types.rst]: Done now.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Does it make sense to squash?

**Comment by ekpyron** [comment]
I squashed at least a bit.

**Review by chriseth — APPROVED** [review]

**Pull request merged**